graph/src/util/lfu_cache.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+impl<K: Ord + Eq + Hash, V> Default for LfuCache<K, V> {
+    fn default() -> Self {
+        LfuCache {
+            queue: PriorityQueue::new(),
+            total_weight: 0,
+            stale_counter: 0,
+        }
+    }
+}
+
+impl<K: Clone + Ord + Eq + Hash + Debug, V: CacheWeight + Default> LfuCache<K, V> {
+    pub fn new() -> Self {
+        LfuCache {
+            queue: PriorityQueue::new(),
+            total_weight: 0,
+            stale_counter: 0,
+        }
+    }
+
+    /// Updates and bumps freceny if already present.
+    pub fn insert(&mut self, key: K, value: V) {
+        let weight = value.weight();
+        match self.get_mut(key.clone()) {
+            None => {
+                self.total_weight += weight;
+                self.queue.push(
+                    CacheEntry {
+                        weight,
+                        key,
+                        value,
+                        will_stale: false,
+                    },
+                    (false, Reverse(1)),
+                );
+            }
+            Some(entry) => {
+                entry.weight = weight;
+                entry.value = value;
+                entry.will_stale = false;
+                self.total_weight += weight - entry.weight;
+            }
+        }
+    }
+
+    fn get_mut(&mut self, key: K) -> Option<&mut CacheEntry<K, V>> {
+        // Increment the frequency by 1
+        let key_entry = CacheEntry::cache_key(key);
+        self.queue
+            .change_priority_by(&key_entry, |(s, Reverse(f))| (s, Reverse(f + 1)));
+        self.queue.get_mut(&key_entry).map(|x| x.0)
+    }
+
+    pub fn get(&mut self, key: &K) -> Option<&V> {
+        self.get_mut(key.clone()).map(|x| &x.value)
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        // `PriorityQueue` doesn't have a remove method, so emulate that by setting the priority to
+        // the absolute minimum and popping.
+        let key_entry = CacheEntry::cache_key(key.clone());
+        self.queue
+            .change_priority(&key_entry, (true, Reverse(u64::min_value())))
+            .and_then(|_| {
+                self.queue.pop().map(|(e, _)| {
+                    assert_eq!(e.key, key_entry.key);
+                    self.total_weight -= e.weight;
+                    e.value
+                })
+            })
+    }
+
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.queue
+            .get(&CacheEntry::cache_key(key.clone()))
+            .is_some()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+
+    pub fn evict(&mut self, max_weight: u64) {
+        if self.total_weight <= max_weight {
+            return;
+        }
+
+        self.stale_counter += 1;
+        if self.stale_counter == STALE_PERIOD {
+            self.stale_counter = 0;
+
+            // Entries marked `will_stale` are effectively set as stale.
+            // All entries are then marked `will_stale`.
```

Comment:
I understand how it works but the comment is confusing. How about something like
```
// Entries marked `will_stale` were not updated in this period. Properly mark them
// as stale in their priorities. Also mark all entities as `will_stale` for the _next_ period
// so that they will be marked stale next time unless they are updated between now
// and then. 
```

About the stale behavior itself: Entries that are not updated in a period become stale. Do we care if they are updated at all? If they are read a lot but never written to, I'd assume we want them to remain cached.

Reply:
That is a great comment, I've used it.

It was a mistake for `get` to not reset the staleness! I've fixed that now.